crates/kubelet/src/device_plugin_manager/mod.rs

```diff
@@ -0,0 +1,954 @@
+        let devices_to_allocate: Vec<String> = available_devices[..quantity].to_vec();
+        // ??: is there a cleaner way to map `allocated_devices.insert(dev.clone())` from bool to
+        // () so can remove {} block:
+        devices_to_allocate.iter().for_each(|dev| {
+            allocated_devices.insert(dev.clone());
+        });
+        allocated_devices_map.insert(resource_name.to_string(), allocated_devices);
+
+        Ok(devices_to_allocate)
+    }
+}
+
+/// Returns the device IDs of all healthy devices that have yet to be allocated.
+fn get_available_devices(
+    devices: &EndpointDevicesMap,
+    allocated_devices: &HashSet<String>,
+) -> Vec<String> {
+    let healthy_devices = devices
+        .iter()
+        .filter_map(|(dev_id, dev)| {
+            if dev.health == HEALTHY {
+                Some(dev_id.clone())
+            } else {
+                None
+            }
+        })
+        .collect::<HashSet<String>>();
+    healthy_devices
+        .difference(allocated_devices)
+        .into_iter()
+        .cloned()
+        .collect::<Vec<String>>()
+}
+
+/// Hosts the device plugin `Registration` service (defined in the device plugin API) for a `DeviceManager`.
+/// Upon device plugin registration, reaches out to its `DeviceManager` to validate the device plugin
+/// and establish a connection with it.
+#[derive(Clone)]
+pub struct DeviceRegistry {
+    device_manager: Arc<DeviceManager>,
+}
+
+impl DeviceRegistry {
+    /// Returns a new `DeviceRegistry` with a reference to a `DeviceManager` to handle each registered device plugin.
+    pub fn new(device_manager: Arc<DeviceManager>) -> Self {
+        DeviceRegistry { device_manager }
+    }
+}
+#[async_trait::async_trait]
+
+impl Registration for DeviceRegistry {
+    async fn register(
+        &self,
+        request: tonic::Request<RegisterRequest>,
+    ) -> Result<tonic::Response<Empty>, tonic::Status> {
+        let register_request = request.get_ref();
+        // Validate
+        self.device_manager
+            .validate(register_request)
+            .await
+            .map_err(|e| tonic::Status::new(tonic::Code::InvalidArgument, format!("{}", e)))?;
+        // Create a list and watch connection with the device plugin
+        // TODO: should the manager keep track of threads?
+        self.device_manager
+            .create_endpoint(register_request)
+            .await
+            .map_err(|e| tonic::Status::new(tonic::Code::NotFound, format!("{}", e)))?;
+        Ok(tonic::Response::new(Empty {}))
+    }
+}
+
+/// Starts the `DeviceManager` by running its `NodePatcher` and serving the `DeviceRegistry` which hosts the device plugin manager's `Registration` service on the socket
+/// specified in the `DeviceManager`.
+/// Returns an error if either the `NodePatcher` or `DeviceRegistry` error.
+pub async fn serve_device_registry(device_manager: Arc<DeviceManager>) -> anyhow::Result<()> {
```

Comment:
Nit: I'd recommend this be a method on `DeviceManager` called `start`. Because you are spawning the tasks, you can then just keep track of those handles in the `DeviceManager` struct. To be clear, definitely not required as this is _exactly_ what the plugin registrar is doing, but it just occurred to me while reading it